types/node/index.d.ts

```diff
@@ -1198,8 +1198,14 @@ declare module "http" {
         createConnection?: (options: ClientRequestArgs, oncreate: (err: Error, socket: net.Socket) => void) => net.Socket;
     }
 
+    interface ServerOptions {
+        IncomingMessage: any;
```

Comment:
I think `typeof IncomingMessage` and `typeof ServerResponse` should work here.

Reply:
But I fear this is not a complete solution as the types of parameters of `requestListener` should be the same as that one passed as options defaulting to `IncomingMessage`/`ServerResponse`. I guess this is possible via generics but I'm not an expert in this. Any ideas are welcome.

Reply:
@Flarna you mean something like this?

```js
class Server<RequestClass, ResponseClass> extends net.Server {
        constructor(requestListener?: (req: RequestClass, res: ResponseClass) => void);
        constructor(options: ServerOptions, requestListener?: (req: RequestClass, res: ResponseClass) => void);

        setTimeout(msecs?: number, callback?: () => void): this;
        setTimeout(callback: () => void): this;
        maxHeadersCount: number;
        timeout: number;
        keepAliveTimeout: number;
    }
```

Reply:
Currently having it like this and compiles. 

```js
interface ServerOptions {
        IncomingMessage: typeof IncomingMessage,
        ServerResponse: typeof ServerResponse
    }

    class Server<RequestClass=IncomingMessage, ResponseClass=ServerResponse> extends net.Server {
        constructor(requestListener?: (req: RequestClass, res: ResponseClass) => void);
        constructor(options: ServerOptions, requestListener?: (req: RequestClass, res: ResponseClass) => void);

        setTimeout(msecs?: number, callback?: () => void): this;
        setTimeout(callback: () => void): this;
        maxHeadersCount: number;
        timeout: number;
        keepAliveTimeout: number;
    }
```

In theory, ServerOptions accepts any subtype of the classes, and the types of the request listener correspond with the given type arguments. It needs improvements, but it compiles and give some decente intellisense.

Reply:
I think this won't transpile as node type definitions here shall work with ancient typescript 2.1 but defaults for generics require 2.3.
I tried the following but this requires even 2.8 and I have the feeling it could be simpler:
```
interface IncomingMessageConstructor {
	new(socket: net.Socket): IncomingMessage;
}

interface ServerResponseConstructor {
	new(req: IncomingMessage): ServerResponse;
}

interface ServerOptions<IM extends IncomingMessageConstructor = typeof IncomingMessage, SR extends ServerResponseConstructor = typeof ServerResponse> {
	IncomingMessage?: IM;
	ServerResponse?: SR;
}

class Server<IM extends IncomingMessageConstructor = IncomingMessageConstructor,
			 SR extends ServerResponseConstructor = ServerResponseConstructor> extends net.Server {
	constructor(options: ServerOptions<IM, SR>, requestListener?: (req: InstanceType<IM>, res: SR) => void);


function createServer<IM extends IncomingMessageConstructor = typeof IncomingMessage,
					  SR extends ServerResponseConstructor = typeof ServerResponse>(options: ServerOptions<IM, SR>,
					  requestListener?: (request: InstanceType<IM>, response: InstanceType<SR>) => void): Server<IM, SR>;
```